### PR TITLE
Implements Targets Page in the React application

### PR DESCRIPTION
**Pull request by sharmamohit123**
Porting Targets page to react using react-bootstrap, along with APIs calling(for backened) using react-sagas.

### CODE REVIEW

File: owtf/webapp/src/containers/TargetsPage/TargetsTable.js [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    }
+
+    let searchTextArray = [];
+    if (multiColumnSearch) {
+      searchTextArray = text.split(' ');
+    } else {
+      searchTextArray.push(text);
+    }
+
+    const result = this.props.targets.filter((product) => {
```
Comment (saganshul) [inline owtf/webapp/src/containers/TargetsPage/TargetsTable.js]: @sharmamohit123 Can you please add some comments here? I really don't want to read whole code to understand what the variable is or functions is like **results**. Also, if you are getting data from backend in undesired format, you can change it.

File: owtf/webapp/src/containers/TargetsPage/TargetsTable.js [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+import { Row, Col, Button, ButtonGroup, Glyphicon, ControlLabel, FormGroup, Label } from 'react-bootstrap';
+import { BootstrapTable, TableHeaderColumn, search } from 'react-bootstrap-table';
+import {connect} from "react-redux";
+import { ClipLoader } from 'react-spinners';
+import './style.scss';
+import { createStructuredSelector } from "reselect";
+import FormControl from 'react-bootstrap/es/FormControl';
+import { changeTarget, deleteTarget, removeTargetFromSession } from './actions';
+import { makeSelectDeleteError, makeSelectRemoveError } from "./selectors";
+import '../../../node_modules/react-bootstrap-table/dist/react-bootstrap-table.min.css'
```
Comment (viyatb) [inline owtf/webapp/src/containers/TargetsPage/TargetsTable.js]: Is this correct? Shouldn't you link `.scss` file for the bootstrap table?

File: owtf/webapp/src/containers/TargetsPage/index.js [hunk 3]
```diff
@@ -1,19 +1,268 @@
...
+            <Row className="add-target-btn">
+              <Button bsStyle="primary" disabled={this.state.disabled} onClick={this.addNewTargets}>
+                Add Targets
+              </Button>
+            </Row>
+          </Col>
+          <Col xs={12} sm={12} md={6} lg={6}>
+            <Row>
+              <Col xs={6} md={6}>
+                <Sessions ref="sessionReference" />
```
Comment (saganshul) [inline owtf/webapp/src/containers/TargetsPage/index.js]: @sharmamohit123 I don't think so you need to refs for this case. You can selector to getCurrentSession. That's all the reason for using store.

File: owtf/webapp/src/containers/Sessions/index.js [hunk 4]
```diff
@@ -71,6 +71,7 @@ export class Sessions extends React.Component {
         <FormGroup>
           <InputGroup>
             <FormControl type="text" placeholder={currentSession.name} readOnly />
+            <FormControl type="hidden" defaultValue={currentSession.id} />
```
Comment (saganshul) [inline owtf/webapp/src/containers/Sessions/index.js]: @sharmamohit123 After you use selectors to get session you don't need this.

File: owtf/webapp/src/containers/Sessions/index.js [hunk 5]
```diff
@@ -128,4 +129,4 @@ const mapDispatchToProps = (dispatch) => {
   };
 };
 
-export default connect(mapStateToProps, mapDispatchToProps)(Sessions);
+export default connect(mapStateToProps, mapDispatchToProps, null, {withRef: true})(Sessions);
```
Comment (saganshul) [inline owtf/webapp/src/containers/Sessions/index.js]: @sharmamohit123 Same here.

File: owtf/webapp/src/containers/TargetsPage/TargetsTable.js [hunk 6]
```diff
@@ -0,0 +1,448 @@
...
+            this.props.handleAlertMsg("warning", base_message + status.failed.length + " attempts failed.");
+        } else {
+          this.props.handleAlertMsg("success", base_message);
+        }
+      }
+    }
+  }
+
+  // function to delete a single target
+  handleRemoveTargetFromSession(target_id) {
```
Comment (saganshul) [inline owtf/webapp/src/containers/TargetsPage/TargetsTable.js]: @sharmamohit123 Don't you think we can utilize `handleRemoveTargetsFromSession` to delete single target too?

File: owtf/webapp/src/containers/TargetsPage/TargetsTable.js [hunk 7]
```diff
@@ -0,0 +1,448 @@
...
+    onChangeTarget: (target) => dispatch(changeTarget(target)),
+    onDeleteTarget: (target_id) => dispatch(deleteTarget(target_id)), 
+    onRemoveTargetFromSession: (session, target_id) => dispatch(removeTargetFromSession(session, target_id)), 
+  };
+}
+
+export default connect(mapStateToProps, mapDispatchToProps)(TargetsTable);
+
+//customize search box
+class MySearchPanel extends React.Component {
```
Comment (saganshul) [inline owtf/webapp/src/containers/TargetsPage/TargetsTable.js]: @sharmamohit123 Move this to `components`.

File: owtf/webapp/src/containers/TargetsPage/TargetsTable.js [hunk 8]
```diff
@@ -0,0 +1,448 @@
...
+  render() {
+    return (
+      <Col>
+        { this.props.searchField }
+      </Col>
+    );
+  }
+}
+
+class RemotePaging extends React.Component {
```
Comment (saganshul) [inline owtf/webapp/src/containers/TargetsPage/TargetsTable.js]: @sharmamohit123 Move this to `components`

File: owtf/webapp/src/containers/TargetsPage/index.js [hunk 9]
```diff
@@ -1,19 +1,268 @@
...
+    this.setState({ disabled: false });
+  }
+
+  componentDidMount() {
+    this.props.onFetchTarget();
+  }
+
+  //return current active session
+  getCurrentSession() {
+    return this.refs.sessionReference.getWrappedInstance().getCurrentSession()
```
Comment (saganshul) [inline owtf/webapp/src/containers/TargetsPage/index.js]: @sharmamohit123 Utilize `selectors`

File: owtf/webapp/src/containers/TargetsPage/style.scss [hunk 10]
```diff
@@ -0,0 +1,28 @@
+.add-target-btn {
+  margin-top: 10px;
+}
+
+/* Spinner*/
```
Comment (saganshul) [inline owtf/webapp/src/containers/TargetsPage/style.scss]: @sharmamohit123 You are gonna use this spinner at lot of place. Move this scss to global scss.

**Comment by saganshul** [comment]
@sharmamohit123 Any component that is not using the store or redux operation and you think we can use the same components at many places, put that in components.

Also, try to use states as immutable. You can utilize immutable.js that is already present in the base. But why? Because it helps when you do testing. Immutability ensures defensive copying and enabling advanced memoization and change detection techniques with simple logic.

Rest, I think is good. Great work BTW :)

File: owtf/webapp/src/containers/TargetsPage/index.js [hunk 11]
```diff
@@ -1,19 +1,268 @@
...
+      handleAlertMsg: this.handleAlertMsg,
+    }
     return (
-      <Sessions />
+      <Grid>
+        <Row>
+          <Col>{this.renderAlert()}</Col>
+        </Row>
+        <Row>
+          <Breadcrumb>
```
Comment (saganshul) [inline owtf/webapp/src/containers/TargetsPage/index.js]: @sharmamohit123 Try using react-router to navigate across breadcrumbs. You can google for current implementations.

**Pull request merged**